File: api/dense/solve/solve_getrs.m
function b = solve_getrs(A, b, varargin)
% solve_getrs  Solves a linear system with A storing the LU factorization
%     b = solve_getrs(A, b) performs back and forward solve on A, where
%     A stores the LU factors in the format returned by LAPACK's DGETRF.
%
%     b = solve_getrs(A, b, n) uses n instead of size(b,1) as n.
%
%     b = solve_getrs(A, b, n, offset) uses vector b(1+offset:n+offset)
%     instead of b.

%#codegen -args {m2c_mat, m2c_vec}
%#codegen solve_getrs_3args -args {m2c_mat, m2c_vec, int32(0)}
%#codegen solve_getrs_4args -args {m2c_mat, m2c_vec, int32(0), int32(0)}

if nargin < 3
    n = int32(size(b,1));
else
    n = int32(varargin{1});
end

if nargin < 4
    offset = int32(0);
else
    offset = int32(varargin{2});
end

% Compute forward on unit lower triangular part
k = int32(1);
for j = 1:n
    k = k + j;
    for i = j+1 : n
        b(i+offset) = b(i+offset) - A(k) * b(j+offset);
        k = k + 1;
    end
end

% Compute backsolve on upper triangular part
k = n * n;
for j = n:-1:1
    b(j+offset) = b(j+offset) / A(k);
    
    for i = j-1:-1:1
        k = k - 1;
        b(i+offset) = b(i+offset) - A(k) * b(j+offset);
    end
    k = k - (n - j) - 2;
end

function test  %#ok<DEFNU>
%!test
%! for k=1:100
%!     A = rand(10,10); b = A * ones(10, 1);
%!
%!     if condest(A)<1.e5
%!         [L, U, p] = lu(A, 'vector');
%!         B = tril(L, -1) + U;
%!     
%!         x = solve_getrs(B, b(p));
%!         assert(norm(A*x - b, inf) <= 1.e-12);
%!     end
%! end

File: api/dense/solve/codegen/lib/solve_getrs/solve_getrs.c
#include "solve_getrs.h"
#include "solve_getrs_types.h"

void solve_getrs(const emxArray_real_T *A, emxArray_real_T *b)
{
  int b_i;
  int i;
  int j;
  int k;
  int n;
  n = b->size[0];
  k = 0;
  for (j = 0; j < n; j++) {
    k = (k + j) + 1;
    i = j + 2;
    for (b_i = i; b_i <= n; b_i++) {
      b->data[b_i - 1] -= A->data[((k + b_i) - j) - 2] * b->data[j];
    }
    k = ((k + n) - j) - 1;
  }
  k = n * n - 1;
  for (j = n; j >= 1; j--) {
    b->data[j - 1] /= A->data[k];
    i = j - 1;
    for (b_i = i; b_i >= 1; b_i--) {
      b->data[b_i - 1] -= A->data[(k + b_i) - j] * b->data[j - 1];
    }
    k = (k - j) + 1;
    k = ((k - n) + j) - 2;
  }
}

void solve_getrs_3args(const emxArray_real_T *A, emxArray_real_T *b, int n)
{
  int b_i;
  int i;
  int j;
  int k;
  k = 0;
  for (j = 0; j < n; j++) {
    k = (k + j) + 1;
    i = j + 2;
    for (b_i = i; b_i <= n; b_i++) {
      b->data[b_i - 1] -= A->data[((k + b_i) - j) - 2] * b->data[j];
    }
    k = ((k + n) - j) - 1;
  }
  k = n * n - 1;
  for (j = n; j >= 1; j--) {
    b->data[j - 1] /= A->data[k];
    i = j - 1;
    for (b_i = i; b_i >= 1; b_i--) {
      b->data[b_i - 1] -= A->data[(k + b_i) - j] * b->data[j - 1];
    }
    k = (k - j) + 1;
    k = ((k - n) + j) - 2;
  }
}

void solve_getrs_4args(const emxArray_real_T *A, emxArray_real_T *b, int n,
                       int offset)
{
  int b_i;
  int i;
  int i1;
  int i2;
  int j;
  int k;
  k = 0;
  for (j = 0; j < n; j++) {
    k = (k + j) + 1;
    i = j + 2;
    for (b_i = i; b_i <= n; b_i++) {
      i1 = (b_i + offset) - 1;
      b->data[i1] -= A->data[((k + b_i) - j) - 2] * b->data[j + offset];
    }
    k = ((k + n) - j) - 1;
  }
  k = n * n - 1;
  for (j = n; j >= 1; j--) {
    i = (j + offset) - 1;
    b->data[i] /= A->data[k];
    i1 = j - 1;
    for (b_i = i1; b_i >= 1; b_i--) {
      i2 = (b_i + offset) - 1;
      b->data[i2] -= A->data[(k + b_i) - j] * b->data[i];
    }
    k = (k - j) + 1;
    k = ((k - n) + j) - 2;
  }
}

void solve_getrs_initialize(void)
{
}

void solve_getrs_terminate(void)
{
}
